crates/chia-datalayer/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+        metadata: &NodeMetadata,
+        index: TreeIndex,
+        blob: [u8; DATA_SIZE],
+    ) -> Result<Self, String> {
+        // TODO: add Err results
+        Ok(Self {
+            parent: Self::parent_from_bytes(&blob)?,
+            index,
+            hash: <[u8; 32]>::try_from(&blob[HASH_RANGE]).unwrap(),
+            specific: match metadata.node_type {
+                NodeType::Internal => NodeSpecific::Internal {
+                    left: TreeIndex::from_be_bytes(<[u8; 4]>::try_from(&blob[LEFT_RANGE]).unwrap()),
+                    right: TreeIndex::from_be_bytes(
+                        <[u8; 4]>::try_from(&blob[RIGHT_RANGE]).unwrap(),
+                    ),
+                },
+                NodeType::Leaf => NodeSpecific::Leaf {
+                    // TODO: this try from really right?
+                    key_value: KvId::from_be_bytes(
+                        <[u8; 8]>::try_from(&blob[KEY_VALUE_RANGE]).unwrap(),
+                    ),
+                },
+            },
+        })
+    }
+
+    fn parent_from_bytes(blob: &[u8; DATA_SIZE]) -> Result<Parent, String> {
+        // TODO: a little setup here for pre-optimization to allow walking parents without processing entire nodes
+        let parent_integer = TreeIndex::from_be_bytes(
+            <[u8; 4]>::try_from(&blob[PARENT_RANGE])
+                .map_err(|e| format!("data blob wrong size: {e}"))?,
+        );
+        match parent_integer {
+            NULL_PARENT => Ok(None),
+            _ => Ok(Some(parent_integer)),
+        }
+    }
+    pub fn to_bytes(&self) -> [u8; DATA_SIZE] {
+        let mut blob: Vec<u8> = Vec::new();
+        match self {
+            Node {
+                parent,
+                specific: NodeSpecific::Internal { left, right },
+                hash,
+                index: _,
+            } => {
+                let parent_integer = match parent {
+                    None => NULL_PARENT,
+                    Some(parent) => *parent,
+                };
+                blob.extend(parent_integer.to_be_bytes());
+                blob.extend(left.to_be_bytes());
+                blob.extend(right.to_be_bytes());
+                blob.extend(hash);
+            }
+            Node {
+                parent,
+                specific: NodeSpecific::Leaf { key_value },
+                hash,
+                index: _,
+            } => {
+                let parent_integer = match parent {
+                    None => NULL_PARENT,
+                    Some(parent) => *parent,
+                };
+                blob.extend(parent_integer.to_be_bytes());
+                blob.extend(key_value.to_be_bytes());
+                blob.extend(hash);
+            }
+        }
+
+        blob.try_into().unwrap()
+    }
+
+    // TODO: yes i know i'm trying to write this code in a non-rusty way and i need to stop that
+    pub fn key_value(&self) -> KvId {
+        let NodeSpecific::Leaf { key_value } = self.specific else {
```

Comment:
then you have a pile of silliness (more of these panic'ing methods) to get at the common field like the parent and hash.  i started that way and then decided to shift this way.  there were a few other options considered along the way that might still be good to think about.  in this form i do still pattern match, just against this specific bit rather than the whole node.